chore: Added CI for publishing agent as Azure site extension

mrickard opened the pull request:
## Description

This adds a GHA workflow for publishing the agent as an Azure Site Extension (Windows support only). This contains both a GitHub Actions workflow, and a set of files in `cloud-tooling/azure-site-extension` with which I've tested. The workflow requires these:

* the `cloud-tooling/azure-site-extension/NewRelic.Azure.WebSites.Extension.NodeAgent.nuspec` file
* the `cloud-tooling/azure-site-extension/Content` directory. (Anything in this directory will be published.)

Necessary repository environment variables for publishing:
* NUGET_API_KEY
* NUGET_SOURCE (the NuGet endpoint for publishing; the .NET and Java teams would know this)

The installation/uninstallation scripts have been brought over from #https://github.com/newrelic/node-newrelic/pull/2448. 

## How to Test

Once it's merged to `main`, this workflow can be triggered manually.

## Related Issues

Closes #2441 
Closes NR-301036

bizob2828 commented on .github/workflows/azure-site-extension.yml:
```diff
@@ -0,0 +1,80 @@
+
+    strategy:
+      matrix:
+        node-version: ['20.x'] # Node 22 isn't yet available on Azure
```
```suggestion
        node-version: ['lts/*'] # Node 22 isn't yet available on Azure
```

mrickard replied:
18.x and 20.x are available for Windows in Azure, so I was targeting those.

bizob2828 commented on .github/workflows/azure-site-extension.yml:
```diff
@@ -0,0 +1,78 @@
+      - name: Install agent
+        working-directory: cloud-tooling/azure-site-extension/Content
+        run: |
+          npm i --prefix Content newrelic@${{ env.AGENT_VERSION }}
```
```suggestion
          npm i --prefix . newrelic@${{ env.AGENT_VERSION }}
```

This ends up creating a subfolder in Content called content

bizob2828 commented on .github/workflows/azure-site-extension.yml:
```diff
@@ -0,0 +1,78 @@
+        working-directory: cloud-tooling/azure-site-extension
+        run: nuget pack "${{ env.PACKAGE_FILENAME }}.nuspec"
+
+      - name: Archive package for verification
```
i'd put a comment on this explaining this is purely for us checking out what's going to get published

bizob2828 commented on .github/workflows/azure-site-extension.yml:
```diff
@@ -0,0 +1,78 @@
+
+    strategy:
+      matrix:
+        node-version: ['18.x', '20.x'] # Node 22 isn't yet available for Windows apps in Azure
```
can you remove `.x`? this ends up putting that in the package name which is very confusing

bizob2828 replied:
also im not certain we need a site extension per node version. we have 2 native libraries: `@newrelic/native-metricss` and `@contrast/fn-inspect` and they provide prebuilts for every supported node version

bizob2828 replied:
turns out our native metrics module works across versions but the `@contrast/fn-inspect` does not.  it only has prebuilt for win32-x64 not ia32 which we need apparently

mrickard replied:
@bizob2828 I don't know if we need ia32--looking at the VM options in Azure, they allow x86-64 and ARM64. I can add a matrix to publish for those two arch options.

mrickard replied:
Azure does let users provision VMs for ARM, but not for Windows as the OS. As site extensions aren't (yet) supported for Linux, we don't have to publish for ARM. 

In the case of 32-vs-64...the default for deploying into Azure is x86-64. x86-32 can be provisioned in Azure, but it's not an easy process: https://learn.microsoft.com/en-us/troubleshoot/azure/virtual-machines/windows/support-32-bit-operating-systems-virtual-machines. As such, 32-bit seems like an edge case.

jsumners-nr replied:
I would completely ignore x86 in this day and age. Very few OS vendors are continuing to directly support it.

bizob2828 replied:
Then we need to triage why the prebuilts aren't working for @contrast/fn-inspect. I was at loss when I looked Friday

jsumners-nr replied:
I'm not sure if I was clear. I'm saying all of the major operating systems are only shipping 64-bit builds when the target platform is Intel. That is, Windows x64 or Linux amd64. We included x86 for Windows in `@newrelic/native-metrics` because people are able to deploy on systems that manage themselves and possibly still utilize Windows x86.

mrickard replied:
After testing, we've determined that building for x86 doesn't give us any advantage: x86 applications will still fall back to function name for CLM, due to the lack of an x86 build of `fn-inspect`. In our tests, we've found an x64-built site extension to instrument equally well for x86 and x64 in every respect other than CLM.

bizob2828 commented on .github/workflows/azure-site-extension.yml:
```diff
@@ -0,0 +1,80 @@
+
+  create_extension_bundle:
+    runs-on: windows-latest
+    needs:
+      - setup_tools
```
Yea i don't understand this as well

mrickard replied:
While I was working this out, I'd wanted to have explicit checks for what's available in the Windows runners vs what needs to be installed. I can fold this into the main job.

mrickard commented on .github/workflows/azure-site-extension.yml:
```diff
@@ -0,0 +1,79 @@
+
+    strategy:
+      matrix:
+        node-version: ['lts/*']
```
It's possible to deploy Node 18.x applications in Azure. Did we want to publish only for LTS?

bizob2828 replied:
yes. our agent is node version agnostic. however, we may have to publish for both x86 and x64, [see](https://github.com/newrelic/node-newrelic/pull/2488#discussion_r1729341609)

jsumners-nr reviewed (approved):
Looks good to me.

bizob2828 commented on .github/workflows/azure-site-extension.yml:
```diff
@@ -0,0 +1,71 @@
+    strategy:
+      matrix:
+        node-version: ['lts/*']
+        arch: [ x64 ]
```
we're only running this on x64 but the nuspec includes this in the name. is this necessary? are we going to document what doesn't work on 32 bit?

mrickard replied:
I'm in favor of documenting the difference *and* removing the architecture from the nuspec name. I'll update.

mrickard replied:
I've restored the x86 build to this, and added some documentation on what doesn't work on 32. (If we supply x86 builds, CLM works as expected on Node 20 for win 32, but not for Node 18.)

mrickard commented:
Given that Node 18 win32 does not have CLM support when we build for x86, and rather than risk customer confusion by building for all the possible combinations of Node version and processor architecture, I've updated this to build for x64 and lts. The readme is now updated to note that full CLM isn't available for win32.

bizob2828 commented on .github/workflows/azure-site-extension.yml:
```diff
@@ -0,0 +1,71 @@
+
+      - name: Set package filename
+        run: |
+          echo "PACKAGE_FILENAME=NewRelic.Azure.WebSites.Extension.NodeAgent-${{env.AGENT_VERSION}}-${{matrix.arch}}" | Out-File -FilePath $env:GITHUB_ENV -Append
```
this still has the arch in the name

mrickard replied:
Whoops. Corrected now.

bizob2828 reviewed (approved):
I can't really test but LGTM
